third_party/tonic/logging/dart_error.cc

```diff
@@ -10,18 +10,66 @@
 namespace tonic {
 namespace DartError {
 const char kInvalidArgument[] = "Invalid argument.";
+DartPersistentValue on_error;
```

Comment:
Is it safe to have DartPersistentValues as global variables?  Should we instead attach this to the dart state?  How does this handle multiple engines?

Reply:
In order to support multiple engines there would need to be a separate `on_error` handler for each engine instance/isolate group.

I'd be interested in asking the Dart team about the safety of the current implementation within a single engine.
The `on_error` closure will be set from the main isolate but will be invoked from whatever isolate is current during the call to `CheckAndHandleError`. 
The invocation itself should work given that both isolates will be in the same isolate group and the invoked closure will be part of the group's shared heap.

But I'd like to understand whether there is any risk in calling the same closure from multiple isolates (and possibly multiple concurrent threads) given that the `on_error` handler calls user-provided code that could do arbitrary mutation of data.

Reply:
@mkustermann @a-siva @mraleph might be able to help here.

Reply:
> But I'd like to understand whether there is any risk in calling the same closure from multiple isolates (and possibly multiple concurrent threads) given that the on_error handler calls user-provided code that could do arbitrary mutation of data.

Also, someone could be setting the handler on the UI thread while a different isolate is accessing it on on secondary isolate, no?

Reply:
I wrote a test that I was sure would fail this (in the latest commit), and to my surprise it passed!

Here's what's up: what looks like a process (or maybe VM)-wide singleton is actually per-isolate. DartPersistentValue keeps a pointer to the isolate, and so even though you're accessing a single `DartPersistentValue`, it's getting that persistetn value _for the current isolate_. This behavior is super weird and I don't want to rely on it, so I've refactored things to more explicitly use `UIDartState::Current` in the reporter callback, and modified that to use the Dart_Handles directly so it all works the way it should.

@jason-simmons: If the current isolate throwing the error tries to access this, it will only get the reporter for its own `PlatformDispatcher`. If it never set one, then exceptions will just get logged. If it did set one, then it will get the one that it set. The new test in shell_unittests.dart verifies this by spawning a second shell and making sure that the onError for each isolate can be used in an interleaved manner.

Reply:
The latest version of this PR is checking `UIDartState::Current()->platform_configuration()` before trying to invoke the user's `on_error` callback.  Only the main isolate has a `PlatformConfiguration`, so unhandled errors from other isolates in the group will instead flow through the `LogUnhandledException` path.

This eliminates the issue described above - it is no longer possible to invoke the `on_error` closure from an isolate other than the one where it originated.

However, users will need to be aware that the `PlatformDispatcher.onError` API only applies to the main isolate.  The `onError` documentation should be updated accordingly.

Reply:
Done. I'm going to mention that `compute` specifically handles this case by forwarding errors to the main isolate.